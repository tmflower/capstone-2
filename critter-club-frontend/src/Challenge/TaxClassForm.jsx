import { useState } from 'react';
import { Paper, FormLabel, RadioGroup, Radio, Typography, Button } from '@mui/material';
import { theme } from '../theme/theme';

/** TaxClassForm renders a question with radio button answers */

export function TaxClassForm({commonName, taxClass, setTaxClass, message, points, setPoints, numQuestions, setNumQuestions  }) {
    
    if (taxClass === "sarcopterygii" || taxClass === "actinopterygii" || taxClass === "chondrichthyes") {
        setTaxClass("fish");
    }

    if (!taxClass) setTaxClass("other");
    
    // Display form with no default answers and no initial message
    const initialState = {taxClass: ''};
    const [formData, setFormData] = useState(initialState);
    const [feedback, setFeedback] = useState('');

    // Deconstruct formData properties
    const { vertebrateGroup } = formData;

    // Set formData to user's input
    const handleChange = (e) => {
        const { name, value } = e.target;
        console.log(e.target.value)
		setFormData((formData) => 
            ({ ...formData, [name]: value })); 
    }

    // compare user's selection to animal data to check if correct; provide corresponding feedback message
    // update user's points
    // increment numQuestions; when numQuestions is >=2, user can submit answers and earn badge
    const handleSubmit = (e) => {
        // compare user's selection to animal data to check if correct; provide corresponding feedback message
        e.preventDefault(); console.log(formData)
        if (vertebrateGroup === taxClass) {
            setFeedback(message.correct);
            setPoints(points+=10);
            setNumQuestions(numQuestions+1);
        }
        else setFeedback(message.incorrect); 
    }

    // clear form responses & message when user clicks "Try again" button
    const handleReset = (e) => {
        setFormData(initialState);
        setFeedback('');
    }

    return (
        <Paper
        elevation={8}
        sx={{ padding: 3, mt: 3 }}>
        <form onSubmit={handleSubmit}>
            <fieldset>
                <RadioGroup>
                <FormLabel sx={{ fontSize: '1.5rem', fontFamily: 'Lexend Deca, Arial',}}>
                    <legend>What taxonomic class does the {commonName.toLowerCase()} belong to?</legend>
                </FormLabel>
                <label htmlFor='Amphibians'>
                    <Radio 
                        // type="radio"
                        name="vertebrateGroup"
                        value="amphibia"
                        id="Amphibians"
                        onChange={handleChange}
                        checked={vertebrateGroup === "amphibia"}
                    />
                    Amphibians
                </label>
                <label htmlFor='Birds'>
                    <Radio 
                        // type="radio"
                        name="vertebrateGroup"
                        value="aves"
                        id="Birds"
                        onChange={handleChange}
                        checked={vertebrateGroup === "aves"}
                        />
                    Birds
                </label>
                <label htmlFor='Mammals'>
                    <Radio
                        // type="radio"
                        name="vertebrateGroup"
                        value="mammalia"
                        id="Mammals"
                        onChange={handleChange}
                        checked={vertebrateGroup === "mammalia"}
                    />
                    {/* </input> */}
                    Mammals
                </label>
                <label htmlFor='Reptiles'>
                    <Radio
                        // type="radio"
                        name="vertebrateGroup"
                        value="reptilia"
                        id="Reptiles"
                        onChange={handleChange}
                        checked={vertebrateGroup === "reptilia"}
                    />
                    {/* </input> */}
                    Reptiles
                </label>
                <label htmlFor='Fish'>
                    <Radio
                        // type="radio"
                        name="vertebrateGroup"
                        value="fish"
                        id="Fish"
                        onChange={handleChange}
                        checked={vertebrateGroup === "fish"}
                    />
                    {/* </input> */}
                    Fish
                </label>
                <label htmlFor='Other'>
                    <Radio 
                        // type="radio"
                        name="vertebrateGroup"
                        value="other"
                        id="Other"
                        onChange={handleChange}
                        checked={vertebrateGroup === "other"}
                    />
                    {/* </input> */}
                    Other
                </label>
                <div>
                    { !feedback ? 
                    <Button type="submit">Check answer</Button>
                    : null }
                    { feedback === message.incorrect ?
                    <Button id="alt-button" type="reset" onClick={handleReset}>Clear answer</Button>
                    : null }
                </div>
                <Typography id="quiz-feedback" variant="h5" sx={{ fontFamily: theme.typography.primary, textAlign: 'center' }}>{feedback}</Typography>
            </RadioGroup>
            </fieldset>
        </form>
        </Paper>
    )
}