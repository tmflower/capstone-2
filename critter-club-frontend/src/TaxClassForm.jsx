import { useState } from 'react';

export function TaxClassForm({commonName, taxClass, setTaxClass, phylum}) {
    
    if (taxClass === "Sarcopterygii" || taxClass === "Actinopterygii" || taxClass === "Chondrichthyes") {
        setTaxClass("Fish");
    }

    if (!taxClass) setTaxClass("Other");
    

    const [formData, setFormData] = useState('');
    const { vertebrateGroup } = formData;

    const handleChange = (e) => {
        const { name, value } = e.target;
		setFormData((formData) => 
            ({ ...formData, [name]: value }));
    }

    const handleSubmit = (e) => {
        // compare user's selection to animal data to check if correct
        // TODO: add functionality for user feedback
        e.preventDefault();
        console.log(vertebrateGroup, taxClass)
        if (vertebrateGroup === taxClass) {
            console.log("YOU ARE CORRECT---class!!! ðŸ˜€ðŸ˜€ðŸ˜€");
        }
        else console.log("SORRY, NOT THIS TIME--class ðŸ˜«ðŸ˜«ðŸ˜«"); 
    }

    const handleReset = (e) => {
        setFormData('');
    }

    return (
        <div>
        <form onSubmit={handleSubmit}>
            <fieldset>
                <legend>What taxonomic class does the {commonName.toLowerCase()} belong to?</legend>
                <label>
                    <input 
                        type="radio"
                        name="vertebrateGroup"
                        value="Amphibia"
                        onChange={handleChange}
                        checked={vertebrateGroup === "Amphibia"}
                        >
                    </input>
                    Amphibians
                </label>
                <label>
                    <input 
                        type="radio"
                        name="vertebrateGroup"
                        value="Aves"
                        onChange={handleChange}
                        checked={vertebrateGroup === "Aves"}
                        >
                    </input>
                    Birds
                </label>
                <label>
                    <input 
                        type="radio"
                        name="vertebrateGroup"
                        value="Mammalia"
                        onChange={handleChange}
                        checked={vertebrateGroup === "Mammalia"}
                        >
                    </input>
                    Mammals
                </label>
                <label>
                    <input 
                        type="radio"
                        name="vertebrateGroup"
                        value="Reptilia"
                        onChange={handleChange}
                        checked={vertebrateGroup === "Reptilia"}
                        >
                    </input>
                    Reptiles
                </label>
                <label>
                    <input 
                        type="radio"
                        name="vertebrateGroup"
                        value="Fish"
                        onChange={handleChange}
                        checked={vertebrateGroup === "Fish"}
                        >
                    </input>
                    Fish
                </label>
                <label>
                    <input 
                        type="radio"
                        name="vertebrateGroup"
                        value="Other"
                        onChange={handleChange}
                        checked={vertebrateGroup === "Other"}
                        >
                    </input>
                    Other
                </label>
                <div>
                    <button type="submit">Check answer</button>
                    <button type="reset" onClick={handleReset}>Clear answer</button>
                </div>
            </fieldset>
        </form>
        </div>
    )
}