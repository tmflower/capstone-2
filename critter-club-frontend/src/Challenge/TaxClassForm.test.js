import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { TaxClassForm } from './TaxClassForm';
import UserContext from '../userContext';

const currentUser = {user: { username: "Tiny", id: 1, password: 'abc123$%^', points: 40, userBadges: ['Koala', 'Jellyfish', 'Painted Turtle'], parentId: 1 }}

const setTaxClass = () => {
    return
}

const taxClass = "mammalia";

const message = {
    correct: "⭐️⭐️⭐️ You got it! ⭐️⭐️⭐️",
    incorrect: "😧😧😧 Sorry, try again. 😧😧😧"
  };

// smoke test
test("renders without crashing", () => {
  
    render(
    <MemoryRouter>
      <UserContext.Provider value={currentUser}>
        <TaxClassForm 
          commonName="Hippopotamus"
          setTaxClass={setTaxClass}
          taxClass={taxClass}
          message={message}/>
      </UserContext.Provider>
    </MemoryRouter>);
  });

// snapshot test
test("matches snapshot", () => {
  const currentUser = {user: { username: "Tiny", id: 1, password: 'abc123$%^', points: 40, userBadges: ['Koala', 'Jellyfish', 'Painted Turtle'], parentId: 1 }}
    const { asFragment } = render(
      <MemoryRouter>
        <UserContext.Provider value={currentUser}>
          <TaxClassForm
            commonName="Hippopotamus"
            setTaxClass={setTaxClass}
            taxClass={taxClass}
            message={message}/>
        </UserContext.Provider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});