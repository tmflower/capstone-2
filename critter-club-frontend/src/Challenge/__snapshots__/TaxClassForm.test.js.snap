// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 css-2e4xct-MuiPaper-root"
  >
    <form>
      <fieldset>
        <div
          class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          role="radiogroup"
        >
          <label
            class="MuiFormLabel-root MuiFormLabel-colorPrimary css-tshmje-MuiFormLabel-root"
          >
            <legend>
              What taxonomic class does the hippopotamus belong to?
            </legend>
          </label>
          <label
            for="Amphibians"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                id="Amphibians"
                name="vertebrateGroup"
                type="radio"
                value="amphibia"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Amphibians
          </label>
          <label
            for="Birds"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                id="Birds"
                name="vertebrateGroup"
                type="radio"
                value="aves"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Birds
          </label>
          <label
            for="Mammals"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                id="Mammals"
                name="vertebrateGroup"
                type="radio"
                value="mammalia"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Mammals
          </label>
          <label
            for="Reptiles"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                id="Reptiles"
                name="vertebrateGroup"
                type="radio"
                value="reptilia"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Reptiles
          </label>
          <label
            for="Fish"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                id="Fish"
                name="vertebrateGroup"
                type="radio"
                value="fish"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Fish
          </label>
          <label
            for="Other"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                id="Other"
                name="vertebrateGroup"
                type="radio"
                value="other"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Other
          </label>
          <div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              Check answer
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-18czui-MuiTypography-root"
            id="quiz-feedback"
          />
        </div>
      </fieldset>
    </form>
  </div>
</DocumentFragment>
`;
